### PR TITLE
Konsti distance metrics

### PR DESCRIPTION
**Pull request by KonstiNik**
In this PR the following main contributions have been made:

- Implementing Wasserstein Distance
- Implementing a parent class for contrastive losses
- Implementing a contrastive loss using isolated potentials
- Implementing a contrastive loss using the InfoNCE loss 
- An example notebook showcasing the usage of the contrastive loss functions

### CODE REVIEW

File: CI/unit_tests/distance_metrics/test_wasserstein_distance.py
Comment (SamTov) [inline CI/unit_tests/distance_metrics/test_wasserstein_distance.py:1]: Typically, if you are using code that is tested somewhere else, like in the scipy package, I suggest not testing it here again. When it comes to shape testing, this is partially important, but at the same time, whether the axis is correct is dependent on how you pass the data in and how it is specified in the call. It doesn't seem necessary to have this test.

File: znnl/loss_functions/__init__.py [hunk 1]
```diff
@@ -41,4 +49,10 @@
     MeanPowerLoss.__name__,
     SimpleLoss.__name__,
     CrossEntropyLoss.__name__,
+    WassersteinLoss.__name__,
```
Comment (SamTov) [inline znnl/loss_functions/__init__.py:52]: Not sure how I feel about this if you can't use it to train a network. I feel like there will be a lot of avoidable issues

File: znnl/loss_functions/contrastive_infonce_loss.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+import jax.numpy as np
+
+from znnl.loss_functions.contrastive_loss import ContrastiveLoss
+
+
+class ContrastiveInfoNCELoss(ContrastiveLoss):
+    """
+    Class for the Contrastive InfoNCE Loss.
+
+    The InfoNCE loss is a type of contrastive loss function.
```
Comment (SamTov) [inline znnl/loss_functions/contrastive_infonce_loss.py:37]: Can you write out the NCE somewhere

File: znnl/loss_functions/contrastive_infonce_loss.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        Parameters
+        ----------
+        temperature : float
+                The temperature parameter for the InfoNCE loss.
+        """
+
+        super().__init__()
+        self.temperature = temperature
+
+    def compute_losses(self, inputs: np.ndarray, targets: np.ndarray):
```
Comment (SamTov) [inline znnl/loss_functions/contrastive_infonce_loss.py:62]: Can you add output type hints here and everywhere

File: znnl/loss_functions/contrastive_infonce_loss.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+        Parameters
+        ----------
+        inputs : np.ndarray
+                The input data of shape (k, n_features), where
+                k = (batch_size^2 - batch_size)/2
+
+        Returns
+        -------
+        dot_product : np.ndarray
```
Comment (SamTov) [inline znnl/loss_functions/contrastive_infonce_loss.py:129]: Returns two things.

File: znnl/loss_functions/exponential_repulsion_loss.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+class ExponentialRepulsionLoss(SimpleLoss):
+    """
+    Class for the exponential repulsion loss.
+
+    The exponential repulsion loss is primarily used for the calculation of the
+    contraction loss as it returns a repulsive measure of the distance between
+    two points.
+    """
+
+    def __init__(self, alpha: float = 0.01, temp: float = 0.1):
```
Comment (SamTov) [inline znnl/loss_functions/exponential_repulsion_loss.py:41]: Can you write out the temperature variable if this is front-facing?

File: znnl/loss_functions/wasserstein_loss.py
Comment (SamTov) [inline znnl/loss_functions/wasserstein_loss.py:1]: If this does remain a loss, it should probably have some big warnings on it so that people don't use it in network training.

**Review by SamTov — CHANGES_REQUESTED** [review]
Looks good. The commenting of the code is great and it reads well. Please check out the comments I made. Be sure to add return hints to all of the functions, they are super helpful. Also, I am worried about the Wasserstein loss being used in training if it will throw an error. From my perspective, this should be either removed as a loss or a warning should appear when users use it.

File: CI/unit_tests/distance_metrics/test_wasserstein_distance.py
Comment (KonstiNik) [reply CI/unit_tests/distance_metrics/test_wasserstein_distance.py:1]: I put that there in case someone might change the method that is used to check whether a new method still works as intended. 
But I can also remove it if you prefer that.

File: znnl/loss_functions/contrastive_infonce_loss.py (hunk 3 above)
Comment (KonstiNik) [reply znnl/loss_functions/contrastive_infonce_loss.py]: Added type hints everywhere in 
[7d57017](https://github.com/zincware/ZnNL/pull/116/commits/7d570172057c05a09394712e0b1ff942ceb5611f)

File: znnl/loss_functions/contrastive_infonce_loss.py (hunk 4 above)
Comment (KonstiNik) [reply znnl/loss_functions/contrastive_infonce_loss.py:129]: Adapt the docstring to return one tuple object in
[7d57017](https://github.com/zincware/ZnNL/pull/116/commits/7d570172057c05a09394712e0b1ff942ceb5611f)

File: znnl/loss_functions/exponential_repulsion_loss.py (hunk 5 above)
Comment (KonstiNik) [reply znnl/loss_functions/exponential_repulsion_loss.py]: spelled temperature out in [60f5a73](https://github.com/zincware/ZnNL/pull/116/commits/60f5a738dd248c199a2622a6f6115a91d1c77a49)

File: CI/unit_tests/distance_metrics/test_wasserstein_distance.py
Comment (SamTov) [reply CI/unit_tests/distance_metrics/test_wasserstein_distance.py:1]: But in this test, the only thing you check is that scipy acts on the correct axis. Is there a way to damage that on the ZnNL side? Or would it be better to have just a shape check test?

File: znnl/loss_functions/wasserstein_loss.py
Comment (SamTov) [reply znnl/loss_functions/wasserstein_loss.py:1]: Can you make this more informative for those who don't full know how NNs work. E.g., 

`cannot be used for gradient calculations, and therefore not for neural network training`

**Comment by SamTov** [comment]
Just the one additional comment about the logging and then it's good to go

File: CI/unit_tests/distance_metrics/test_wasserstein_distance.py
Comment (KonstiNik) [reply CI/unit_tests/distance_metrics/test_wasserstein_distance.py:1]: Yes, this is since the metric is implemented with a map:
https://github.com/zincware/ZnNL/blob/e854df7cc2381f3bfeb69a1caf63b5ece32a3855/znnl/distance_metrics/wasserstein_distance.py#L71

File: znnl/loss_functions/__init__.py (hunk 1 above)
Comment (KonstiNik) [reply znnl/loss_functions/__init__.py:52]: Done in 
[da01395](https://github.com/zincware/ZnNL/pull/116/commits/da01395f4712b503ba946fa02c6bfe4276767f17)

File: znnl/loss_functions/wasserstein_loss.py
Comment (KonstiNik) [reply znnl/loss_functions/wasserstein_loss.py:1]: Done in 
[da01395](https://github.com/zincware/ZnNL/pull/116/commits/da01395f4712b503ba946fa02c6bfe4276767f17)

File: CI/unit_tests/distance_metrics/test_wasserstein_distance.py
Comment (SamTov) [reply CI/unit_tests/distance_metrics/test_wasserstein_distance.py:1]: You're still not testing any ZnNL operation, this would be testing the map operation. But we can leave it in if you would like.

**Review by SamTov — APPROVED** [review]
Great addition to the codebase!
